### DISCUSSION TITLE
Criteria API の Case 条件に Domain Class を指定する場合について

### DISCUSSION THREAD

**Discussion by ironclad-15256304** [domaframework/doma#999] [Q&A] [answered]
### 相談内容

いつも大変お世話になっております。

Criteria API を使用した SQL の構築時、 `Expressions.when` を使用することで、 select 句中に case 式を埋め込むことが可能であると認識しています。
こちらの機能を使用して下記のような SQL を構築してみたところ、型の不一致が原因でコンパイルエラーとなります。

```
Table00_ table = new Table00_();
new NativeSql(Config.get(this))
        .from(table)
        .select(
                Expressions.when(c -> {
                    c.eq(table.type, Type.About, Expressions.literal("概要")); // Type は Enum Domain Class
                }, Expressions.literal("詳細"))
        )
        .fetch();
```

where 句内で同様のこと（ `eq` での比較）を行おうとした場合にはコンパイルエラーが起きないのですが、 `Expressions.when` 中の `eq` ではエラーとなります。
一応、ドメインクラスが適用されていない Entity を作成し、そちらを利用すれば良いものではあると認識していますが、可能であればそのままドメインクラスを使用して実装したいと考えております。

こちらを回避し、ドメインクラスが適用されている列に対して case 式を記述する方法はございますでしょうか。もしくは、何かしらこちらの実装に問題があるのでしょうか。
お忙しいところ大変恐縮ですが、実現方法や対応策等ご教示頂けますと幸いです。

**Comment by nakamura-to** [comment, marked as answer]
現状、CASE式に値をバインドできないです。これまでそういうケースを想定していませんでした。
今後修正を検討したいと思います。

代替策としてあまり良い方法はないですね。Criteria APIではなくSQLテンプレートを使っていただくのが一番良いと思います。

**Reply by ironclad-15256304** [reply]
なるほどですね。 SQL Template やアプリケーション側での吸収含め、検討してみようと思います。
迅速なご回答有難うございます！

**Reply by nakamura-to** [reply]
Doma 2.54.2で対応しました。
https://github.com/domaframework/doma/releases/tag/2.54.2

**Reply by ironclad-15256304** [reply]
返信が遅れ申し訳ありません。毎回迅速なご対応をいただき、有難うございます。
対応していただいたバージョンを取り込み次第、実装に組み込んでいこうと思います！
